File: bugsnag-android-core/src/main/java/com/bugsnag/android/Stacktrace.java
package com.bugsnag.android;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Serialize an exception stacktrace and mark frames as "in-project"
 * where appropriate.
 */
class Stacktrace implements JsonStream.Streamable {

    private static final int STACKTRACE_TRIM_LENGTH = 200;

    private final List<Map<String, Object>> trace;

    Stacktrace(StackTraceElement[] stacktrace, Collection<String> projectPackages) {
        this.trace = serializeStacktrace(stacktrace, projectPackages);
    }

    Stacktrace(List<Map<String, Object>> frames) {
        if (frames.size() >= STACKTRACE_TRIM_LENGTH) {
            this.trace = frames.subList(0, STACKTRACE_TRIM_LENGTH);
        } else {
            this.trace = frames;
        }
    }

    @Override
    public void toStream(@NonNull JsonStream writer) throws IOException {
        writer.beginArray();
        for (Map<String, Object> element : trace) {
            writer.value(element);
        }
        writer.endArray();
    }

    private List<Map<String, Object>> serializeStacktrace(StackTraceElement[] trace,
                                                          Collection<String> projectPackages) {
        List<Map<String, Object>> list = new ArrayList<>();

        for (int k = 0; k < trace.length && k < STACKTRACE_TRIM_LENGTH; k++) {
            StackTraceElement el = trace[k];
            Map<String, Object> frame = serializeStackframe(el, projectPackages);

            if (frame != null) {
                list.add(frame);
            }
        }
        return list;
    }

    @Nullable
    private Map<String, Object> serializeStackframe(StackTraceElement el,
                                                    Collection<String> projectPackages) {
        Map<String, Object> map = new HashMap<>();
        try {
            String methodName;
            if (el.getClassName().length() > 0) {
                methodName = el.getClassName() + "." + el.getMethodName();
            } else {
                methodName =  el.getMethodName();
            }
            map.put("method", methodName);

            String filename = el.getFileName() == null ? "Unknown" : el.getFileName();
            map.put("file", filename);
            map.put("lineNumber", el.getLineNumber());

            if (inProject(el.getClassName(), projectPackages)) {
                map.put("inProject", true);
            }
            return map;
        } catch (Exception lineEx) {
            return null;
        }
    }

    private static boolean inProject(String className, Collection<String> projectPackages) {
        for (String packageName : projectPackages) {
            if (packageName != null && className.startsWith(packageName)) {
                return true;
            }
        }
        return false;
    }
}

File: bugsnag-android-core/src/main/java/com/bugsnag/android/Stacktrace.kt
package com.bugsnag.android

import java.io.IOException
import kotlin.math.min

/**
 * Serialize an exception stacktrace and mark frames as "in-project"
 * where appropriate.
 */
internal class Stacktrace : JsonStream.Streamable {

    companion object {
        private const val STACKTRACE_TRIM_LENGTH = 200

        /**
         * Calculates whether a stackframe is 'in project' or not by checking its class against
         * [Configuration.getProjectPackages].
         *
         * For example if the projectPackages included 'com.example', then
         * the `com.example.Foo` class would be considered in project, but `org.example.Bar` would
         * not.
         */
        fun inProject(className: String, projectPackages: Collection<String>): Boolean? {
            return when {
                projectPackages.any { className.startsWith(it) } -> true
                else -> null
            }
        }

        fun serializeStackframe(
            el: StackTraceElement,
            projectPackages: Collection<String>,
            logger: Logger
        ): Stackframe? {
            try {
                val className = el.className
                val methodName = when {
                    className.isNotEmpty() -> className + "." + el.methodName
                    else -> el.methodName
                }

                return Stackframe(
                    methodName,
                    el.fileName ?: "Unknown",
                    el.lineNumber,
                    inProject(className, projectPackages)
                )
            } catch (lineEx: Exception) {
                logger.w("Failed to serialize stacktrace", lineEx)
                return null
            }
        }
    }

    val trace: MutableList<Stackframe>

    constructor(frames: MutableList<Stackframe>) {
        trace = limitTraceLength(frames)
    }

    constructor(
        stacktrace: Array<StackTraceElement>,
        projectPackages: Collection<String>,
        logger: Logger
    ) {
        // avoid allocating new subLists or Arrays by only copying the required number of frames
        // mapping them to our internal Stackframes as we go, roughly equivalent to
        // stacktrace.take(STACKTRACE_TRIM_LENGTH).mapNotNullTo(ArrayList()) { ... }
        val frameCount = min(STACKTRACE_TRIM_LENGTH, stacktrace.size)
        trace = ArrayList(frameCount)
        for (i in 0 until frameCount) {
            val frame = serializeStackframe(stacktrace[i], projectPackages, logger)
            if (frame != null) {
                trace.add(frame)
            }
        }
    }

    private fun limitTraceLength(frames: MutableList<Stackframe>): MutableList<Stackframe> {
        return when {
            frames.size >= STACKTRACE_TRIM_LENGTH -> frames.subList(0, STACKTRACE_TRIM_LENGTH)
            else -> frames
        }
    }

    @Throws(IOException::class)
    override fun toStream(writer: JsonStream) {
        writer.beginArray()
        trace.forEach { writer.value(it) }
        writer.endArray()
    }
}
